src/compiler/parser.ts

```diff
@@ -17,11 +17,22 @@ namespace ts {
     let IdentifierConstructor: new (kind: SyntaxKind, pos: number, end: number) => Node;
     let SourceFileConstructor: new (kind: SyntaxKind, pos: number, end: number) => Node;
 
-    interface Fail { __FAIL: void; }
-    /** Only value of type Fail. */
-    const Fail: Fail = { __FAIL: undefined };
-    function isFail(x: {} | null | undefined | void): x is Fail {
-        return x === Fail;
+    interface Fail extends Node { kind: SyntaxKind.Unknown; }
+    function fail(): Fail {
+        return createNode(SyntaxKind.Unknown) as Fail;

```

Comment:
I still think that `fail` and `failList` should be singleton values local to the `Parser` namespace. You can easily initialize them for the first time in `initializeState` so that we don't create excess objects when speculative parsing fails.